### PR TITLE
vpc: fix ips on wrong interfaces after rebooting vpc vrs

### PR DESCRIPTION
**Pull request by ustcweizhou**
## Description

When vpc tiers are added/removed and public ips in different range is added/removed, the nics in VPC VRs are added/removed, it leads to inconsecutive device in VPC VRs.
When we reboot the VR, (or VR is rebooted by some other reasons like HA), the ips in VPC VRs might be assigned to wrong interfaces.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

File: server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java [hunk 1]
```diff
@@ -834,4 +837,11 @@ public boolean postStateTransitionEvent(final StateMachine2.Transition<State, Vi
         // once from VpcVirtualNetworkApplianceManagerImpl and once from VirtualNetworkApplianceManagerImpl itself
         return true;
     }
+
+    private Nic getNicWithUpdatedDeviceId(final long nicId, int deviceId) {
+        NicVO nic = _nicDao.findById(nicId);
+        nic.setDeviceId(deviceId);
+        _nicDao.update(nic.getId(), nic);
+        return nic;
+    }
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java:846]: method name is not clear, Either define & use separate methods (for "updating nic device id" and "getting nic with device id"), or change the name appropriately.

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java:846]: `getNicAndUpdtaeDeviceId()` might have been more accurate but this is clearly stating what it does on the tin (to me) , Can you be more explicit about what you want @sureshanaparti ?

Comment (sureshanaparti) [reply server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java:846]: > `getNicAndUpdtaeDeviceId()` might have been more accurate but this is clearly stating what it does on the tin (to me) , Can you be more explicit about what you want @sureshanaparti ?

simply "_updateNicWithDeviceId()_" returning nic would be ok.

Comment (weizhouapache) [reply server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java:846]: @sureshanaparti `updateNicWithDeviceId()` looks good to me. @DaanHoogland what's your opinion ?

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java:846]: yes, fine. the only problem i have is with my own spelling; `getNicAndUpdtaeDeviceId()` misses a 'k' ;)

Comment (weizhouapache) [reply server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java:846]: @sureshanaparti @DaanHoogland thanks guys. code is changed.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]

**Comment by weizhouapache** [comment]
@DaanHoogland please do not merge this pr. I found an issue which might be caused by this pr.

**Comment by DaanHoogland** [comment]
@weizhouapache converted to draft so it is clear 4.15 is not waiting for it. please change back if you disagree

**Comment by weizhouapache** [comment]
@DaanHoogland fixed a bug by commit https://github.com/apache/cloudstack/pull/4467/commits/f6038166c5c1a46fe383a9ee059ae397d948baff

in description of commit 7988f51ac07782ad9b9873c32438fd7cdb24edba, it says that "Force a restart of keepalived if conntrackd is not running or configuration has changed". however, in code, it is "and" not "or"

**Review by DaanHoogland — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@weizhouapache can you add soem test data/results?

**Comment by rhtyd** [comment]
@blueorangutan package

**Review by sureshanaparti — APPROVED** [review]
changes LGTM, haven't tested.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Needs a round of test against vmware
@blueorangutan test centos7 vmware-67u3

**Comment by DaanHoogland** [comment]
I do not see how these template tests (all pertaining to checksums are related. We may have uncovered an unrelated issue thought it does not happen in other vmware tests. @rhtyd , you want to make the call?

**Pull request merged**